src/ast/ast_validations.c

```diff
@@ -1299,356 +1509,272 @@ static AST_Validation _ValidateClauseOrder(const AST *ast) {
-		_AST_GetWithReferences(node, identifiers);
-	} else {
-		_AST_GetIdentifiers(node, identifiers);
-	}
+	return !ErrorCtx_EncounteredError() ? AST_VALID : AST_INVALID;
 }
 
-static AST_Validation _Validate_Aliases_DefinedInClause(const cypher_astnode_t *clause,
-														rax *defined_aliases) {
-	AST_Validation res = AST_VALID;
-	rax *referred_identifiers = raxNew();
-
-	// Get defined identifiers.
-	_AST_GetDefinedIdentifiers(clause, defined_aliases);
-
-	// Get referred identifiers.
-	_AST_GetReferredIdentifiers(clause, referred_identifiers);
-
-	raxIterator it;
-	_prepareIterateAll(referred_identifiers, &it);
-
-	// See that each referred identifier is defined.
-	while(raxNext(&it)) {
-		int len = it.key_len;
-		unsigned char *alias = it.key;
-		if(raxFind(defined_aliases, alias, len) == raxNotFound) {
-			ErrorCtx_SetError("%.*s not defined", len, alias);
-			res = AST_INVALID;
-			break;
-		}
-	}
+static VISITOR_STRATEGY _default_visit
+(
+	const cypher_astnode_t *n,
+	bool start,
+	ast_visitor *visitor
+) {
+	ASSERT(n != NULL);
 
-	// Clean up:
-	raxStop(&it);
-	raxFree(referred_identifiers);
-	return res;
+	return VISITOR_RECURSE;
 }
 
-/* Check that all referred identifiers been defined. */
-static AST_Validation _Validate_Aliases_Defined(const AST *ast) {
-	AST_Validation res = AST_VALID;
+// void AST_MappingInit() {
+// build the global mapping from ast-node-type to visiting function
+bool AST_ValidationsMappingInit() {
+	// create a mapping for the validations
+	validations_mapping = AST_Visitor_mapping_new();
 
-	// Retrieve the indices of each WITH clause to properly set the bounds of each scope.
-	// If the query does not have a WITH clause, there is only one scope.
-	uint end_offset = cypher_ast_query_nclauses(ast->root);
-	rax *defined_aliases = raxNew();
-	for(uint i = 0; i < end_offset; i++) {
-		const cypher_astnode_t *clause = cypher_ast_query_get_clause(ast->root, i);
-		// For each clause, confirm that all referred aliases have been previously defined.
-		res = _Validate_Aliases_DefinedInClause(clause, defined_aliases);
-		if(res != AST_VALID) break;
-		if(cypher_astnode_type(clause) == CYPHER_AST_WITH) {
-			// each WITH clause marks the beginning of a new scope for defined aliases
-			// if the WITH clause contains a star projection, all variables from
-			// the previous scope are carried over
-			if(!cypher_ast_with_has_include_existing(clause)) {
-				raxFree(defined_aliases);
-				defined_aliases = raxNew();
-			}
-			_AST_GetDefinedIdentifiers(clause, defined_aliases);
-		}
+	// initialize all entires with the default visit function
+	for(uint i = 0; i < 256; i++) {
+		AST_Visitor_mapping_register(validations_mapping, i, _default_visit);
 	}
-	raxFree(defined_aliases);
-	return res;
+	
+	// populate the mapping with validation functions
+
+	// supported types
+	AST_Visitor_mapping_register(validations_mapping, CYPHER_AST_MATCH, _Validate_MATCH_Clause);
```

Comment:
As a general comment, please review each visit function and see if we can move checks that are able to "exit" to the `start` phase for early and cheaper execution.